File: scripts/Testing/Python/08_test_call_subgraph.py
#!/usr/bin/env python3

from __future__ import annotations

import importlib.util
import sys
import tempfile
from pathlib import Path

import pandas as pd


REPO_ROOT = Path(__file__).resolve().parents[3]
SUBGRAPH_SCRIPT = REPO_ROOT / "scripts" / "Internals" / "Python" / "08_call_subgraph.py"
GG_NODES = REPO_ROOT / "data" / "processed" / "gene_gene_graph_nodes.tsv.gz"
GG_EDGES = REPO_ROOT / "data" / "processed" / "gene_gene_graph_edges.tsv.gz"


def load_subgraph_module():
    spec = importlib.util.spec_from_file_location("conseguiR_step8", SUBGRAPH_SCRIPT)
    module = importlib.util.module_from_spec(spec)
    assert spec.loader is not None
    sys.modules[spec.name] = module
    spec.loader.exec_module(module)
    return module


def ortools_available(module) -> bool:
    return getattr(module, "cp_model", None) is not None


def make_diffusion_fixture(tmpdir: Path, n_genes: int = 80) -> Path:
    nodes = pd.read_csv(GG_NODES, sep="\t", low_memory=False)
    edges = pd.read_csv(GG_EDGES, sep="\t", low_memory=False)

    edge_node_ids = pd.unique(pd.concat([edges["from"], edges["to"]], ignore_index=True))
    candidate_ids = list(edge_node_ids[:n_genes])

    node_lookup = nodes.set_index("node_id")
    candidate_ids = [node_id for node_id in candidate_ids if node_id in node_lookup.index]

    diffusion = pd.DataFrame(
        {
            "node_id": candidate_ids,
            "gene_name": [node_lookup.loc[node_id, "name"] for node_id in candidate_ids],
            "post_germline": [5.0 - (idx * 0.03) for idx in range(len(candidate_ids))],
            "post_somatic": [3.5 - (idx * 0.02) for idx in range(len(candidate_ids))],
            "post_epigenomic": [2.5 - (idx * 0.015) for idx in range(len(candidate_ids))],
            "prediff_norm": [1.0 + (idx * 0.01) for idx in range(len(candidate_ids))],
            "rank_shift": [0.0 for _ in candidate_ids],
        }
    )
    diffusion["post_norm"] = (
        diffusion["post_germline"] ** 2
        + diffusion["post_somatic"] ** 2
        + diffusion["post_epigenomic"] ** 2
    ) ** 0.5

    diffusion_path = tmpdir / "gene_reg_graph_diffusion_all_genes.tsv"
    diffusion.to_csv(diffusion_path, sep="\t", index=False)
    return diffusion_path


def test_run_cardinality_subgraph_calling_live():
    module = load_subgraph_module()

    if not ortools_available(module):
        print(
            "Skipping live step 8 optimization test because OR-Tools is not installed. "
            "Run this test from `lymphoma_graph_env` to exercise the solver path."
        )
        return

    with tempfile.TemporaryDirectory(prefix="conseguiR_step8_test_") as tmp:
        tmpdir = Path(tmp)
        diffusion_path = make_diffusion_fixture(tmpdir, n_genes=80)

        config = module.SubgraphConfig(
            diffusion_path=str(diffusion_path),
            gg_nodes_path=str(GG_NODES),
            gg_edges_path=str(GG_EDGES),
            output_dir=str(tmpdir),
            output_stem="gene_gene_selected_subgraph_test",
            target_genes=12,
            candidate_pool_size=40,
            max_edges_in_model=500,
            max_time_seconds=30,
            num_workers=4,
        )

        result = module.run_cardinality_subgraph_calling(config=config)
        selected_nodes = result["selected_nodes"]
        selected_edges = result["selected_edges"]

        assert selected_nodes.shape[0] == 12, "Selected node cardinality does not match the request."
        assert {"node_id", "gene_name", "prize"}.issubset(selected_nodes.columns)
        assert {"gene_u", "gene_v"}.issubset(selected_edges.columns)
        assert Path(result["output_paths"]["nodes_path"]).exists()
        assert Path(result["output_paths"]["edges_path"]).exists()
        assert Path(result["output_paths"]["summary_path"]).exists()
        assert Path(result["output_paths"]["graphml_path"]).exists()

        print("Selected genes from step 8 test fixture:")
        print(selected_nodes.head(20).to_string(index=False))


def test_validate_diffusion_results_negative_missing_columns():
    module = load_subgraph_module()
    bad = pd.DataFrame({"node_id": ["TP53"], "post_norm": [1.0]})

    try:
        module.validate_diffusion_results(bad, prize_column="post_norm")
    except ValueError as exc:
        assert "missing columns" in str(exc)
    else:
        raise AssertionError("Expected diffusion validation to fail for missing gene_name.")


def test_run_cardinality_subgraph_calling_negative_missing_ortools():
    module = load_subgraph_module()
    if ortools_available(module):
        return

    with tempfile.TemporaryDirectory(prefix="conseguiR_step8_no_ortools_") as tmp:
        tmpdir = Path(tmp)
        diffusion_path = make_diffusion_fixture(tmpdir, n_genes=40)

        config = module.SubgraphConfig(
            diffusion_path=str(diffusion_path),
            gg_nodes_path=str(GG_NODES),
            gg_edges_path=str(GG_EDGES),
            output_dir=str(tmpdir),
            target_genes=10,
            candidate_pool_size=20,
        )

        try:
            module.run_cardinality_subgraph_calling(config=config)
        except ImportError as exc:
            assert "OR-Tools is required" in str(exc)
        else:
            raise AssertionError("Expected step 8 to fail clearly when OR-Tools is unavailable.")


def test_run_cardinality_subgraph_calling_negative_no_overlap():
    module = load_subgraph_module()

    with tempfile.TemporaryDirectory(prefix="conseguiR_step8_bad_overlap_") as tmp:
        tmpdir = Path(tmp)
        diffusion = pd.DataFrame(
            {
                "node_id": [f"NOT_A_REAL_GENE_{i}" for i in range(20)],
                "gene_name": [f"NOT_A_REAL_GENE_{i}" for i in range(20)],
                "post_norm": [1.0 + i for i in range(20)],
            }
        )
        diffusion_path = tmpdir / "bad_diffusion.tsv"
        diffusion.to_csv(diffusion_path, sep="\t", index=False)

        config = module.SubgraphConfig(
            diffusion_path=str(diffusion_path),
            gg_nodes_path=str(GG_NODES),
            gg_edges_path=str(GG_EDGES),
            output_dir=str(tmpdir),
            target_genes=10,
            candidate_pool_size=20,
        )

        try:
            module.run_cardinality_subgraph_calling(config=config)
        except ValueError as exc:
            assert "could be matched" in str(exc)
        else:
            raise AssertionError("Expected step 8 to fail when diffusion genes do not overlap the graph.")


def test_run_cardinality_subgraph_calling_negative_bad_edges():
    module = load_subgraph_module()

    with tempfile.TemporaryDirectory(prefix="conseguiR_step8_bad_edges_") as tmp:
        tmpdir = Path(tmp)
        diffusion_path = make_diffusion_fixture(tmpdir, n_genes=40)

        bad_edges = pd.read_csv(GG_EDGES, sep="\t", low_memory=False).drop(columns=["confidence"])
        bad_edges_path = tmpdir / "bad_edges.tsv"
        bad_edges.to_csv(bad_edges_path, sep="\t", index=False)

        config = module.SubgraphConfig(
            diffusion_path=str(diffusion_path),
            gg_nodes_path=str(GG_NODES),
            gg_edges_path=str(bad_edges_path),
            output_dir=str(tmpdir),
            target_genes=10,
            candidate_pool_size=20,
        )

        try:
            module.run_cardinality_subgraph_calling(config=config)
        except ValueError as exc:
            assert "missing columns" in str(exc)
        else:
            raise AssertionError("Expected step 8 to fail for malformed gene-gene edges.")


def test_run_cardinality_subgraph_calling_negative_small_candidate_pool():
    module = load_subgraph_module()

    with tempfile.TemporaryDirectory(prefix="conseguiR_step8_bad_pool_") as tmp:
        tmpdir = Path(tmp)
        diffusion_path = make_diffusion_fixture(tmpdir, n_genes=30)

        config = module.SubgraphConfig(
            diffusion_path=str(diffusion_path),
            gg_nodes_path=str(GG_NODES),
            gg_edges_path=str(GG_EDGES),
            output_dir=str(tmpdir),
            target_genes=15,
            candidate_pool_size=10,
        )

        try:
            module.run_cardinality_subgraph_calling(config=config)
        except ValueError as exc:
            assert "at least as large" in str(exc)
        else:
            raise AssertionError("Expected step 8 to fail when candidate_pool_size < target_genes.")


def main():
    test_run_cardinality_subgraph_calling_live()
    test_validate_diffusion_results_negative_missing_columns()
    test_run_cardinality_subgraph_calling_negative_missing_ortools()
    test_run_cardinality_subgraph_calling_negative_no_overlap()
    test_run_cardinality_subgraph_calling_negative_bad_edges()
    test_run_cardinality_subgraph_calling_negative_small_candidate_pool()
    print("Step 8 subgraph-calling tests passed.")


if __name__ == "__main__":
    main()

File: scripts/Testing/R/08_test_call_subgraph.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(data.table)
  library(testthat)
  devtools::load_all(".", quiet = TRUE)
})

source("scripts/Internals/R/06b_python_basilisk.R")
source("scripts/Internals/R/08_call_subgraph.R")
source("R/backend_resources.R")
default_step8_test_output_dir <- "data/processed/test_outputs/step8"

make_step8_test_path <- function(stem, ext = "") {
  dir.create(default_step8_test_output_dir, recursive = TRUE, showWarnings = FALSE)
  file.path(
    default_step8_test_output_dir,
    paste0(stem, "_", format(Sys.time(), "%Y%m%d%H%M%S"), "_", sprintf("%06d", sample.int(999999L, 1L)), ext)
  )
}

python_runtime_candidates <- function() {
  candidates <- c("python", "python3")
  candidates[nzchar(Sys.which(candidates))]
}

python_module_available <- function(module_name) {
  candidates <- python_runtime_candidates()
  if (length(candidates) == 0L) {
    return(FALSE)
  }

  for (python_cmd in candidates) {
    status <- system2(
      python_cmd,
      args = c("-c", shQuote(paste0("import ", module_name))),
      stdout = TRUE,
      stderr = TRUE
    )
    exit_status <- attr(status, "status")
    if (is.null(exit_status) || identical(exit_status, 0L)) {
      return(TRUE)
    }
  }
  FALSE
}

make_diffusion_fixture <- function() {
  dt <- data.table(
    node_id = c("G1", "G2", "G3"),
    gene_name = c("G1", "G2", "G3"),
    post_norm = c(1.0, 1.5, 0.8),
    post_germline = c(0.8, 1.0, 0.5),
    post_somatic = c(0.2, 0.3, 0.2),
    post_epigenomic = c(0.5, 0.6, 0.4)
  )
  dt[, post_vulnerability := sqrt(
    pmax(post_germline, 0)^2 +
      pmax(post_somatic, 0)^2 +
      abs(post_epigenomic)^2
  )]
  dt[, post_integrated := (post_germline + post_somatic + post_epigenomic) / sqrt(3)]
  dt
}

make_gene_gene_graph_fixture <- function() {
  list(
    nodes = data.table(
      node_id = c("G1", "G2", "G3"),
      node_type = c("gene", "gene", "gene")
    ),
    edges = data.table(
      from = c("G1", "G2", "G2"),
      to = c("G2", "G3", "G1"),
      confidence = c(0.9, 0.8, 0.7),
      weight = c(1.0, 1.0, 1.0)
    )
  )
}

python_ortools_available <- function() {
  candidates <- python_runtime_candidates()
  if (length(candidates) == 0L) {
    return(FALSE)
  }

  for (python_cmd in candidates) {
    status <- system2(
      python_cmd,
      args = c("-c", shQuote("import ortools")),
      stdout = TRUE,
      stderr = TRUE
    )
    exit_status <- attr(status, "status")
    if (is.null(exit_status) || identical(exit_status, 0L)) {
      return(TRUE)
    }
  }
  FALSE
}

write_subgraph_fixture_files <- function() {
  diffusion <- make_diffusion_fixture()
  gg <- make_gene_gene_graph_fixture()
  output_prefix <- make_step8_test_path("subgraph_fixture")
  diffusion_path <- paste0(output_prefix, "_diffusion.tsv")
  gg_nodes_path <- paste0(output_prefix, "_gg_nodes.tsv")
  gg_edges_path <- paste0(output_prefix, "_gg_edges.tsv")
  fwrite(diffusion, diffusion_path, sep = "\t")
  fwrite(gg$nodes, gg_nodes_path, sep = "\t")
  fwrite(gg$edges, gg_edges_path, sep = "\t")
  list(
    diffusion_path = diffusion_path,
    gg_nodes_path = gg_nodes_path,
    gg_edges_path = gg_edges_path
  )
}

test_validate_diffusion_results <- function() {
  diffusion <- make_diffusion_fixture()
  validated <- validate_diffusion_results(diffusion, prize_column = "post_integrated")

  expect_true(is.data.table(validated))
  expect_equal(nrow(validated), 3L)
}

test_validate_diffusion_results_negative <- function() {
  bad_diffusion <- data.table(node_id = "G1", post_norm = 1.0)
  expect_error(
    validate_diffusion_results(bad_diffusion, prize_column = "post_norm"),
    regexp = "Diffusion results are missing required columns",
    fixed = TRUE
  )
}

test_validate_gene_gene_graph_inputs <- function() {
  fixture <- make_gene_gene_graph_fixture()
  validated_nodes <- validate_gene_gene_nodes(fixture$nodes)
  validated_edges <- validate_gene_gene_edges(fixture$edges)

  expect_true(is.data.table(validated_nodes))
  expect_true(is.data.table(validated_edges))
}

test_validate_gene_gene_graph_inputs_negative <- function() {
  bad_nodes <- data.table(node_id = "G1", node_type = "protein")
  expect_error(
    validate_gene_gene_nodes(bad_nodes),
    regexp = "Gene-gene node table contains unsupported node_type values",
    fixed = TRUE
  )

  bad_edges <- data.table(from = "G1", to = "G2", weight = 1.0)
  expect_error(
    validate_gene_gene_edges(bad_edges),
    regexp = "Gene-gene edge table is missing required columns",
    fixed = TRUE
  )
}

test_run_cardinality_subgraph_calling_integration <- function() {
  if (length(python_runtime_candidates()) == 0L) {
    skip("Python is not available on PATH.")
  }
  if (!python_module_available("numpy") || !python_module_available("pandas")) {
    skip("Required Python packages are not installed for step 8 integration testing.")
  }
  if (!python_ortools_available()) {
    skip("OR-Tools is not available for live step 8 testing.")
  }

  fixture_paths <- write_subgraph_fixture_files()
  output_dir <- make_step8_test_path("conseguiR_step8")
  result <- run_cardinality_subgraph_calling(
    diffusion_path = fixture_paths$diffusion_path,
    gg_nodes_path = fixture_paths$gg_nodes_path,
    gg_edges_path = fixture_paths$gg_edges_path,
    output_dir = output_dir,
    output_stem = "gene_gene_selected_subgraph_test",
    target_genes = 2L,
    candidate_pool_size = 3L,
    max_edges_in_model = 50L,
    max_time_seconds = 30L,
    num_workers = 1L
  )

  expect_true(file.exists(result$output_paths$nodes_path))
  expect_true(file.exists(result$output_paths$edges_path))
  expect_true(file.exists(result$output_paths$summary_path))
  expect_true(file.exists(result$output_paths$graphml_path))
  nodes_dt <- fread(result$output_paths$nodes_path)
  expect_equal(nrow(nodes_dt), 2L)
}

main <- function() {
  test_that("diffusion result validation works", {
    test_validate_diffusion_results()
  })

  test_that("diffusion validation reports missing required columns", {
    test_validate_diffusion_results_negative()
  })

  test_that("gene-gene graph validation works", {
    test_validate_gene_gene_graph_inputs()
  })

  test_that("gene-gene graph validation rejects malformed input", {
    test_validate_gene_gene_graph_inputs_negative()
  })

  test_that("step 8 Python-backed subgraph wrapper runs when OR-Tools is available", {
    test_run_cardinality_subgraph_calling_integration()
  })
}

if (sys.nframe() == 0) {
  main()
}
